Collections: poll collection until publication job is finished

tw4l opened the pull request:
Initial implementation of #973 

Updated screenshot of status indicator:

![image](https://github.com/webrecorder/browsertrix-cloud/assets/6758804/8d61c784-685b-40a0-9a01-2da73825dfdb)

Ideally we'd want to use the sync upload callback to load a progress bar, but that will require some additional refactoring. Progress on this has been slower than expected because every time I try to publish a large collection my dev environment breaks (I think running out of memory due to the way the download stream is currently implemented).

tw4l committed:
Poll collection detail until publication task finished

- Add published boolean field to Collection
- Modify publishing endpoint to return URL early and complete in background
task
- Poll from frontend until published is true
- Show loading spinner while publication is still in progress
- Disable delete and publish buttons while publication is in progress

SuaYoo commented on frontend/src/pages/org/collection-detail.ts:
```diff
@@ -123,7 +137,7 @@ export class CollectionDetail extends LiteElement {
   >
       <p class="text-left">
-        Embed this published collection in other site using the following embed code and ReplayWeb.page:
+        Embed this published collection in another website using the following embed code and ReplayWeb.page:
```
FYI I refactored this section https://github.com/webrecorder/browsertrix-cloud/pull/974 so this line will create a merge conflict, I've went ahead and fixed the copy [here in my branch](https://github.com/webrecorder/browsertrix-cloud/pull/974/commits/47071b666662766c4446ddd15199fc9d65a9aeb9) if you want to revert it here!

tw4l replied:
will do, ty!

tw4l committed:
Remove print statement

SuaYoo commented on frontend/src/pages/org/collection-detail.ts:
```diff
@@ -60,8 +65,14 @@ export class CollectionDetail extends LiteElement {
           ${this.collection?.name || html`<sl-skeleton></sl-skeleton>`}
         </h2>
+        ${when(this.isLoading, () => html`
+          <div class="flex justify-center mr-2 p-2">
```
This should align things vertically:
```suggestion
          <div class="flex items-center justify-center mr-2 p-2">
```

tw4l committed:
Update frontend/src/pages/org/collection-detail.ts

Co-authored-by: sua yoo <sua@webrecorder.org>

tw4l committed:
Revert to using publishedUrl, but don't return immediately

tw4l committed:
Fix coll id reference

tw4l committed:
Frontend fix

tw4l committed:
Allow publishedUrl to be undefined

tw4l committed:
Check if publishedUrl exists before checking length

tw4l committed:
Persist collection publishing status to backend

tw4l committed:
Improve updating of collection with publishing

tw4l committed:
Check for publishing and publishedUrl in status conditionals

tw4l committed:
Disable correct delete button

tw4l committed:
Update publishedUrl to empty string when first publishing

tw4l committed:
Set publishing to false on unpublishing just to be safe

tw4l committed:
Make publishing a bool

ikreymer committed:
bump memory for backend
attempt to keep track of upload size

ikreymer committed:
Merge branch 'publish-public-and-streaming-dl' into publish-poll-detail-until-finished

ikreymer committed:
remove unused

ikreymer committed:
add progress bar!
track upload updates via queue

ikreymer committed:
fix typo

ikreymer committed:
size tweaks

ikreymer committed:
reformat

ikreymer committed:
use fork of stream-zip which supports streaming

tw4l committed:
Reset backend memory to previous values

tw4l commented on backend/btrixcloud/colls.py:
```diff
@@ -467,9 +581,9 @@ class FileLikeObj:
-            return b"".join(
-                up_to_iter(float("inf") if size is None or size < 0 else size)
-            )
+            print("size read", size, flush=True)
```
Can maybe remove this print statement now?

tw4l commented on backend/btrixcloud/colls.py:
```diff
@@ -437,6 +523,25 @@ def member_files():
+        """upload callback"""
+        self.counter += num
+        update_time = time.time()
+        if (update_time - self.last_update) > 1:
```
```suggestion
        if (update_time - self.last_update) > 5:
```

tw4l commented:
@ikreymer Tested locally with backend memory reset to previous limits, and both publication and downloading are working great as proper streams! The progress bar also works nicely, even on navigating away from and then back to the collection page. Also tried downloading the WACZ and opening it in replayweb.page just to be extra sure, and that also worked well.

Code also looks great! Just left a few small comments/suggestions.

tw4l commented:
Very nice work on this!

ikreymer committed:
uploads cleanup, reset current percent when starting an upload

ikreymer committed:
refactor:
- move streaming download/upload to storages
- ensure wacz files are unchanged between downloads/uploads, use fixed timestamp
- fix unpublish
- fix progress update race condition

ikreymer committed:
fix publishing coll access with minio:
- return '/data/' endpoint for publishedUrl
- resolve relative URLs on frontend

ikreymer committed:
fix deletion after return change
